azkaban-common/src/main/java/azkaban/storage/StorageCleaner.java

```diff
@@ -0,0 +1,156 @@
+    this.storage = storage;
+    this.databaseOperator = databaseOperator;
+
+    this.maxArtifactsPerProject = props.getInt(AZKABAN_STORAGE_ARTIFACT_MAX_RETENTION, 0);
+    this.projectLoader = projectLoader;
+    checkArgument(this.maxArtifactsPerProject >= 0,
+        String.format("Invalid value for %s : %d", AZKABAN_STORAGE_ARTIFACT_MAX_RETENTION,
+            this.maxArtifactsPerProject));
+
+    if (isCleanupPermitted()) {
+      log.info(String.format("%s Config: Max %d artifact(s) retained per project",
+          AZKABAN_STORAGE_ARTIFACT_MAX_RETENTION, this.maxArtifactsPerProject));
+    } else {
+      log.warn("Project cleanup disabled. All artifacts will be stored.");
+    }
+  }
+
+  @VisibleForTesting
+  boolean isCleanupPermitted() {
+    return this.maxArtifactsPerProject > 0;
+  }
+
+  /**
+   * Remove all but last N artifacts as configured by AZKABAN_STORAGE_ARTIFACT_MAX_RETENTION
+   *
+   * Since multiple versions can share the same filename, the algo is to collect all filenames and
+   * from them, remove the latest ones. The remaining ones are deleted by the respective storage.
+   *
+   * @param projectId project ID
+   */
+  public void cleanupProjectArtifacts(final int projectId) {
+    if (!isCleanupPermitted()) {
+      return;
+    }
+    final List<VersionResourcePair> pvrList = fetchPvrList(projectId);
+    if (pvrList.size() <= this.maxArtifactsPerProject) {
+      return;
+    }
+    cleanUpStorageArtifacts(projectId, pvrList);
+
+    final int lowestRetainedVersionIndex = this.maxArtifactsPerProject - 1;
+    log.warn(String.format(
+        "Removing the following version entries for project [id: %d]: %s", projectId,
+        pvrList.subList(lowestRetainedVersionIndex, pvrList.size()).stream()
+            .map(pvr -> pvr.version).collect(Collectors.toList()))
+    );
+
+    cleanUpDatabase(projectId, pvrList.get(lowestRetainedVersionIndex).version);
+  }
+
+
+  private void cleanUpStorageArtifacts(final int projectId,
+      final List<VersionResourcePair> pvrList) {
+    final Set<String> allResourceIds = pvrList.stream().map(pvr -> pvr.resourceId)
+        .collect(Collectors.toSet());
+    final Set<String> doNotDeleteSet = pvrList.subList(0, this.maxArtifactsPerProject).stream()
+        .map(pvr -> pvr.resourceId)
+        .collect(Collectors.toSet());
+    allResourceIds.removeAll(doNotDeleteSet);
+
+    log.warn(String.format("Deleting project artifacts [id: %d]: %s", projectId, allResourceIds));
+    allResourceIds.forEach(this.storage::delete);
+  }
+
+  private void cleanUpDatabase(final int projectId, final int lowestRetainedVersion) {
+    this.projectLoader.cleanOlderProjectVersion(projectId, lowestRetainedVersion);
+    try {
+      this.databaseOperator.update(SQL_DELETE_OLDER_VERSIONS, projectId, lowestRetainedVersion);
+    } catch (final SQLException e) {
+      log.error("Error while deleting version entries from DB for project ID: " + projectId, e);
```

Comment:
Even if we face this error, this shouldn't affect the execution of Azkaban. In the next upload, the clean up will be attempted again. The recovery piece was missing and has been added.